modulus/models/module.py

```diff
@@ -65,55 +105,256 @@ def debug(self):
+        init_argspec = inspect.getfullargspec(torch_model_class.__init__)
+        model_argnames = init_argspec.args[1:]  # Exclude 'self'
+        model_defaults = init_argspec.defaults or []
+        defaults_dict = dict(
+            zip(model_argnames[-len(model_defaults) :], model_defaults)
+        )
+
+        # Define the signature of new init
+        params = [inspect.Parameter("self", inspect.Parameter.POSITIONAL_OR_KEYWORD)]
+        params += [
+            inspect.Parameter(
+                argname,
+                inspect.Parameter.POSITIONAL_OR_KEYWORD,
+                default=defaults_dict.get(argname, inspect.Parameter.empty),
+            )
+            for argname in model_argnames
+        ]
+        init_signature = inspect.Signature(params)
+
+        # Replace ModulusModel.__init__ signature with new init signature
+        ModulusModel.__init__.__signature__ = init_signature
+
+        # Generate a unique name for the created class
+        new_class_name = f"{torch_model_class.__name__}ModulusModel"
+        ModulusModel.__name__ = new_class_name
+
+        # Add this class to the dict of dynamically created classes
+        ModulusModel.register(ModulusModel, new_class_name)
+
+        return ModulusModel
+
+    @classmethod
+    def register(cls, model, name: Union[str, None] = None):
+        """
+        Registers a model under a specific name.
+
+        Parameters
+        ----------
+        model
+            The model to be registered.
+        name : Union[str,None], optional
+            The name to register the model under. If none is provided, the model's name
+        """
+
+        # If no name provided, use the model's name
+        if name is None:
+            name = model.__name__
+
+        # Check if name already in use
+        if name in Module._dynamically_created_classes:
+            raise ValueError(f"Name {name} already in use")
+
+        # Add this class to the dict of dynamically created classes
+        Module._dynamically_created_classes[name] = model
+
+    @classmethod
+    def factory(cls, name: str):
+        """
+        Returns a registered model given its name.
+
+        Parameters
+        ----------
+        name : str
+            The name of the registered model.
+
+        Returns
+        -------
+        model
+            The registered model.
+
+        Raises
+        ------
+        KeyError
+            If no model is registered under the provided name.
+        """
+
+        if name in Module._dynamically_created_classes:
+            return Module._dynamically_created_classes[name]
+
+        group = "modulus.models"
+        try:
+            entry_point = entrypoints.get_group_named(group)[name]
+            model_class = entry_point.load()
+            return model_class
+        except KeyError:
+            raise KeyError(f"No model is registered under the name {name}")
+
+    @classmethod
+    def _clear_dynamically_created_classes(cls):
```

Comment:
I feel like this should be in registry. Like registry should be a singleton that gets initialized on import

Reply:
Agreed, suggested the same here: https://github.com/NVIDIA/modulus/pull/65#discussion_r1266295764

Reply:
Hey @akshaysubr and @NickGeneva , Can you please look at this file and tell me if thats more what you had in mind for a singlton class model registry. https://github.com/NVIDIA/modulus/pull/65/files#diff-d0c16db3b5d360b1240b22b92e2e109e2ed837b2b9ed2ad99d02666210cc5e91.

Reply:
@loliverhennigh Yes, this was what I had in mind. I'll add a few comments at specific lines